orchagent/response_publisher.cpp

```diff
@@ -150,22 +210,70 @@ void ResponsePublisher::writeToDB(const std::string &table, const std::string &k
         if (attrs.size())
         {
             applStateTable.set(key, attrs);
-            RecordDBWrite(table, key, attrs, op);
         }
     }
     else if (op == DEL_COMMAND)
     {
         applStateTable.del(key);
-        RecordDBWrite(table, key, {}, op);
     }
 }
 
 void ResponsePublisher::flush()
 {
-    m_pipe->flush();
+    m_ntf_pipe->flush();
+    if (m_update_thread != nullptr)
+    {
+        {
+            std::lock_guard<std::mutex> lock(m_lock);
+            m_queue.push(entry{
+                .table = "",
+                .key = "",
+                .values = std::vector<swss::FieldValueTuple>{},
+                .op = "",
+                .replace = false,
+                .flush = true,
+                .shutdown = false,
+            });
+        }
+        m_signal.notify_one();
+    }
+    else
+    {
+        m_pipe->flush();
+    }
 }
 
 void ResponsePublisher::setBuffered(bool buffered)
 {
     m_buffered = buffered;
 }
+
+void ResponsePublisher::dbUpdateThread()
+{
+    while (true)
+    {
+        entry e;
+        {
+            std::unique_lock<std::mutex> lock(m_lock);
+            while (m_queue.empty())
+            {
+                m_signal.wait(lock);
+            }
+
+            e = m_queue.front();
+            m_queue.pop();
+        }
+        if (e.shutdown)
+        {
+            break;
+        }
+        if (e.flush)
+        {
+            m_pipe->flush();
```

Comment:
>m_pipe [](http://example.com/codeflow?start=12&length=6)

You should not share DBConnector between threads. It is not thread-safe.

Reply:
This pipe object is only used in the "DB update thread". It is not used in the main thread beside constructor.

The DB write operations are completed handled by the "DB update thread" if thread mode is enabled.
(If thread mode is disabled, the main thread will update the DB. There is no "DB update thread" in that case.)